src/NServiceBus.Core/Receiving/ReceiveComponent.cs

```diff
@@ -211,7 +221,67 @@ IPushMessages BuildMessagePump()
             return messagePumpFactory();
         }
 
-        ReceiveConfiguration configuration;
+        static void LoadMessageHandlers(Configuration configuration, List<Type> orderedTypes, IConfigureComponents container, IList<Type> availableTypes)
+        {
+            var types = new List<Type>(availableTypes);
+
+            foreach (var t in orderedTypes)
+            {
+                types.Remove(t);
+            }
+
+            types.InsertRange(0, orderedTypes);
+
+            ConfigureMessageHandlersIn(configuration, types, container);
+        }
+
+        static void ConfigureMessageHandlersIn(Configuration configuration, IEnumerable<Type> types, IConfigureComponents container)
+        {
+            var handlerRegistry = configuration.MessageHandlerRegistry ?? new MessageHandlerRegistry();
+
+            foreach (var t in types.Where(IsMessageHandler))
+            {
+                container.ConfigureComponent(t, DependencyLifecycle.InstancePerUnitOfWork);
+                handlerRegistry.RegisterHandler(t);
+            }
+
+            container.RegisterSingleton(handlerRegistry);
+        }
+
+        public static bool IsMessageHandler(Type type) // TODO: used in LoadMessageHandlersExtensions.ExecuteTheseHandlersFirst() extension method. Should it be moved elsewhere?
```

Comment:
Do you plan to do something here regarding the TODO? perhaps just add it to our POA?